descriptors/web_app.js

```diff
@@ -0,0 +1,174 @@
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+// Espruino devices publish a UART service by default.
+const nordicUARTService = '6e400001-b5a3-f393-e0a9-e50e24dcca9e';
+// The test service defined in device_code.js
+const testServiceUUID = '47ffa036-81c8-11eb-8dcd-0242ac130003';
+// The test characteristic defined in device_code.js
+const testCharacteristicUUID = '47ffa252-81c8-11eb-8dcd-0242ac130003';
+
+let gattServer = undefined;
+
+/**
+ * Load the device code to the Espruino IDE.
+ */
+function loadEspruinoDeviceCode() {
+  fetch('device_code.js').then(response => response.text()).then(data => {
+    let url = 'http://www.espruino.com/webide?code=' + encodeURIComponent(data);
+    window.open(url, '_window');
+  });
+}
+
+function onGattDisconnected(evt) {
+  const device = evt.target;
+  logInfo(`Disconnected from GATT on device ${device.name}.`);
+  assertFalse(gattServer.connected, 'Server connected');
+}
+
+async function startTest() {
+  clearStatus();
+  logInfo('Starting test');
+
+  $('btn_start_test').disabled = true;
+  $('btn_load_code').disabled = true;
+
+  // Verify the characteristic descriptors defened in device_code.js
+  const verifyTestServiceDescriptors = async (service) => {
+    assertEquals(testServiceUUID, service.uuid, 'incorrect service UUID');
+
+    const characteristics = await service.getCharacteristics();
+    assertEquals(1, characteristics.length, 'expected 1 characteristic');
+    for (characteristic of characteristics) {
+      switch (characteristic.uuid) {
+        case testCharacteristicUUID:
+          let descriptors = await characteristic.getDescriptors();
+          assertEquals(1, descriptors.length);
+          const descriptor = await characteristic.getDescriptor(
+            BluetoothUUID.getDescriptor('gatt.characteristic_user_description'));
+          assertEquals(
+            BluetoothUUID.getDescriptor('gatt.characteristic_user_description'),
+            descriptor.uuid);
+          assertEquals(characteristic, descriptor.characteristic,
+            'descriptor characteristic mismatch');
+          assertEquals(null, descriptor.value, 'value should be null');
+          const value = await descriptor.readValue();
+          const decoder = new TextDecoder('utf-8');
+          let strVal = decoder.decode(value);
+          assertEquals('Test UINT32 characteristic', strVal);
+
+          strVal = decoder.decode(value);
+          assertEquals('Test UINT32 characteristic', strVal);
```

Comment:
Duplicate logic. Did you mean to test something else?

Reply:
Yes, I meant `descriptor.value`. thx.